### PR TITLE
David comments

### PR DESCRIPTION
**Pull request by dtnaylor**

### CODE REVIEW

File: mergebot.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        return
+    print(f"pr num: {pr_num}")
+
+    # get PR from github REST API
+    g = Github(os.environ["INPUT_GITHUB_TOKEN"])
+    repo = g.get_repo(os.environ["GITHUB_REPOSITORY"])
+    pr = repo.get_pull(pr_num)
+    print(f"pr title: {pr.title}")
+
+    # My PR was just closed/merged, this could make other PRs dirty, update them
```
Comment (dtnaylor) [inline mergebot.py:67]: Would closing it really impact other PRs?

File: mergebot.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    if event["action"] == "closed":
+        print(
+            f"pr {pr.number} was closed/merged. looking for others sharing this base {pr.base.ref}"
+        )
+        for p in repo.get_pulls(base=pr.base.ref):
+            print(f"found pr: {p.number}, {labeled_and_open(p)}, {p.mergeable_state}")
+            if not labeled_and_open(p) or p.mergeable_state != "behind":
+                continue
+            # Just remove and re-add the label. This will kick off an mergebot
+            # run against the PR.
```
Comment (dtnaylor) [inline mergebot.py:78]: I guess this is the only way to kick it? :( So we'll get a bunch of "bot removed label" "bot added label" spam in the other PR's feed?

File: mergebot.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        pr_num = event["number"]
+    elif (
+        "workflow_run" in event
+        and "pull_requests" in event["workflow_run"]
+        and event["workflow_run"]["pull_requests"]
+    ):
+        pr_num = event["workflow_run"]["pull_requests"][0]["number"]
+    else:
+        print("no pull requests in this event...")
+        return
```
Comment (dtnaylor) [inline mergebot.py:58]: So in the corner case that someone admin-merged a branch (without a PR), that means we won't kick other things in the autorebase queue?

That just means that they'll let tests finish running and _then_ notice that they need to rebase instead of immediately restarting the tests, right?

File: mergebot.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    # check if branch is out-of-date and has conflicting files
+    if pr.mergeable_state == "dirty":
+        give_up(pr, "Conflicting files")
+        return
+
+    # check if branch is out-of-date, if so rebase
+    if pr.mergeable_state == "behind":
+        run(
+            'git config --global user.email "nefeli-mergebot[bot]@users.noreply.github.com"'
```
Comment (dtnaylor) [inline mergebot.py:111]: If we ever really want this to be open source + usable by other people, we probably need a conf file for things like this. (low priority)

File: mergebot.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        run('git config --global user.name "nefeli-mergebot[bot]"')
+        try:
+            print(f"rebasing pr {pr.number}")
+            pr.remove_from_labels(LABEL)
+            run(
+                f'git clone https://x-access-token:{os.environ["INPUT_GITHUB_TOKEN"]}@github.com/{os.environ["GITHUB_REPOSITORY"]}.git'
+            )
+            dir = os.environ["GITHUB_REPOSITORY"].split("/")[1]
+            run(f"cd {dir} && git checkout {pr.head.ref}")
+            run(f"cd {dir} && git rebase {pr.base.ref} --autosquash")
```
Comment (dtnaylor) [inline mergebot.py:122]: Why `--autosquash`? Don't we want the author to need to do that manually (and let jenkins complain if they don't)?

File: mergebot.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    # check if branch is mergeable, if so merge
+    if not pr.mergeable:
+        give_up(pr, "Unknown error when trying to merge, check log")
+        return
+
+    print(f"Merging PR {pr.number}")
+    pr.merge(merge_method="rebase")
+
+    # Mark issue as done in JIRA if we have that set up
+    # TODO(mukerjee): verify with folks if we want this
```
Comment (dtnaylor) [inline mergebot.py:164]: One thought: could we just make two separate labels? `autorebase` and `autorebase-and-close`?

Also, brainstorming possible alternative label names: `merge-it` and `merge-it-and-close`

File: mergebot.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+from jira import JIRA
+
+from github import Github
+
+LABEL = "bug"
+
+ERROR_COMMENT = """<h2><p align="center">:construction: :construction_worker: :fire: Giving up on autorebase, please investigate: :fire: :construction_worker: :construction:</p>
+
+<p align="center">:rotating_light: {} :rotating_light:</p></h2>
+"""
```
Comment (dtnaylor) [inline mergebot.py:18]:
```suggestion
ERROR_COMMENT = ":rotating_light: <b>Giving up on autorebase:</b> {}"
```

**Review by dtnaylor — COMMENTED** [review]
attn @mukerjee

File: mergebot.py (hunk 6 above)
Comment (mukerjee) [reply mergebot.py:164]: ooo good idea. i like the `merge-it` variant

Comment (mukerjee) [reply mergebot.py:164]: code-wise though, it's easier to just make these two separate labels e.g., `merge-it` and `close-it`. (where `close-it` does nothing unless `merge-it` is there as well.)

File: mergebot.py (hunk 5 above)
Comment (mukerjee) [reply mergebot.py:122]: this is current behavior of autorebase as well... no need to force a user to do something that's pretty trivial for no real reason.

File: mergebot.py (hunk 4 above)
Comment (mukerjee) [reply mergebot.py:111]: yeah, good point. we'll get to that later.

File: mergebot.py (hunk 3 above)
Comment (mukerjee) [reply mergebot.py:58]: yeah, i'm not worried about that corner case. we don't let people admin merge on master currently. but yeah, in the worst case, github would just block you from being able to merge and the bot would give up.

File: mergebot.py (hunk 1 above)
Comment (mukerjee) [reply mergebot.py:67]: "close" here also means "merged". this is just how github's API handles a merge... the PR becomes "closed".

File: mergebot.py (hunk 2 above)
Comment (mukerjee) [reply mergebot.py:78]: i have some ideas, like we talked about.

File: mergebot.py (hunk 6 above)
Comment (mukerjee) [reply mergebot.py:164]: nevermind-- going with `merge-it` and `merge-it-and-close`

**Comment by mukerjee** [comment]
thanks for the comments @dtnaylor -- they're on master now.
